File: Plugins/AudioPlayer/src/android/app/src/main/java/com/shortsands/audioplayer/AudioBible.java
package com.shortsands.audioplayer;

import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.PowerManager;
import android.util.Log;
import com.shortsands.aws.AwsS3Cache;
import com.shortsands.aws.CompletionHandler;
import java.io.File;
import java.io.IOException;

/**
 * Created by garygriswold on 8/30/17.
 */

public class AudioBible implements MediaPlayer.OnErrorListener, MediaPlayer.OnCompletionListener,
        MediaPlayer.OnSeekCompleteListener {

    private static final String TAG = "AudioBible";

    private static AudioBible instance = null;
    static AudioBible shared(AudioBibleController controller) {
        if (AudioBible.instance == null) {
            AudioBible.instance = new AudioBible(controller);
        }
        return(AudioBible.instance);
    }

    private final AudioBibleController controller;
    private AudioAnalytics audioAnalytics;
    // Transient Variables
    private AudioReference currReference;
    private AudioReference nextReference;
    private MediaPlayer mediaPlayer;
    private MediaPlayer nextPlayer;

    private AudioBible(AudioBibleController controller) {
        super();
        this.controller = controller;
    }

    MediaPlayer getPlayer() {
        return this.mediaPlayer;
    }

    AudioReference getCurrReference() {
        return this.currReference;
    }

    boolean isPlaying() {
        return (this.mediaPlayer != null) ? this.mediaPlayer.isPlaying() : false;
    }

    void beginReadFile(AudioReference reference) {
        Log.d(TAG, "BibleReader.BEGIN Read File");
        this.currReference = reference;
        this.audioAnalytics = new AudioAnalytics(controller.activity,
            reference.tocAudioBook.testament.bible.mediaSource,
            reference.damId(),
            reference.dbpLanguageCode(),
            reference.textVersion(),
            reference.silLang());
        BeginReadFileCompletion handler = new BeginReadFileCompletion();
        this.readVerseMetaData(reference);
        AudioPlayState.retrieve(this.controller.activity, reference.damId());
        AwsS3Cache.shared().readFile(reference.getS3Bucket(),
                reference.getS3Key(),
                Integer.MAX_VALUE,
                handler);
    }

    class BeginReadFileCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                initAudio(file);
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "BeginReadFile Failed " + exception.toString());
        }
    }

    private void initAudio(File url) {
        if (url != null) {
            this.mediaPlayer = this.initPlayer(url);
            if (this.mediaPlayer != null) {
                long seekTime = AudioPlayState.currentState.position;
                if (seekTime > 100L) {
                    this.mediaPlayer.setOnSeekCompleteListener(this);
                    this.mediaPlayer.seekTo((int) seekTime);
                } else {
                    this.onSeekComplete(this.mediaPlayer);
                }
            }
        } else {
            Log.e(TAG, "URL is null");
        }
    }

    @Override
    public void onSeekComplete(MediaPlayer player) {
        player.setOnSeekCompleteListener(null);
        this.play();
        this.preFetchNextChapter(this.currReference);
        this.controller.playHasStarted(player);
    }

    private MediaPlayer initPlayer(File file) {
        String url = file.getAbsolutePath();
        MediaPlayer player = new MediaPlayer();
        player.setLooping(false);
        player.setAudioStreamType(AudioManager.STREAM_VOICE_CALL); // Replace with AudioAttributes in SDK 21
        try {
            player.setDataSource(url);
            player.prepare();
            player.setOnCompletionListener(this);
            player.setWakeMode(this.controller.activity, PowerManager.PARTIAL_WAKE_LOCK);
            return player;
        } catch(IOException ioe) {
            Log.e(TAG, "Error in AudioBible.initPlayer " + ioe.toString());
            return null;
        }
    }

    void play() {
        Log.d(TAG, "\n*********** PLAY *************");
        if (this.mediaPlayer != null) {
            this.mediaPlayer.start();
            long currTime = this.mediaPlayer.getCurrentPosition();
            this.audioAnalytics.playStarted(this.currReference.toString(), currTime);
        }
    }

    void pause() {
        Log.d(TAG, "\n*********** PAUSE *************");
        if (this.mediaPlayer != null) {
            this.mediaPlayer.pause();
            long currTime = this.mediaPlayer.getCurrentPosition();
            this.audioAnalytics.playEnded(this.currReference.toString(), currTime);
            this.updateMediaPlayStateTime(this.currReference);
        }
    }

    void stop() {
        if (this.mediaPlayer != null && this.mediaPlayer.isPlaying()) {
            this.pause();
        }
        this.controller.playHasStopped();
        if (this.mediaPlayer != null) {
            this.mediaPlayer.reset();
            this.mediaPlayer.release();
            this.mediaPlayer = null;
        }
        if (this.nextPlayer != null) {
            this.nextPlayer.reset();
            this.nextPlayer.release();
            this.nextPlayer = null;
        }
    }

    @Override
    public void onCompletion(MediaPlayer player) {
        this.advanceToNextItem();
    }

    @Override
    public boolean onError(MediaPlayer player, int what, int extra) {
        //this.errorTime = System.currentTimeMillis();
        String message;
        switch (what) {
            case MediaPlayer.MEDIA_ERROR_IO:
                message = "MEDIA ERROR IO";
                break;
            case MediaPlayer.MEDIA_ERROR_MALFORMED:
                message = "MEDIA ERROR MALFORMED";
                break;
            case MediaPlayer.MEDIA_ERROR_NOT_VALID_FOR_PROGRESSIVE_PLAYBACK:
                message = "MEDIA ERROR NOT VALID FOR PROGRESSIVE PLAYBACK";
                break;
            case MediaPlayer.MEDIA_ERROR_SERVER_DIED:
                message = "MEDIA ERROR SERVER DIED";
                break;
            case MediaPlayer.MEDIA_ERROR_TIMED_OUT:
                message = "MEDIA ERROR TIMED OUT";
                break;
            case MediaPlayer.MEDIA_ERROR_UNKNOWN:
                message = "MEDIA ERROR UNKNOWN";
                break;
            case MediaPlayer.MEDIA_ERROR_UNSUPPORTED:
                message = "MEDIA ERROR UNSUPPORTED";
                break;
            default:
                message = "Unknown Error " + what;
        }
        Log.e(TAG, "onError " + message + " " + extra);

        mediaPlayer.reset();
        //this.onErrorRecovery = true;
        //this.progressBar.setVisibility(View.VISIBLE);
        //Uri videoUri = Uri.parse(this.videoUrl);
        //this.videoView.setVideoURI(videoUri);
        return true;
    }

    private void updateMediaPlayStateTime(AudioReference reference) {
        AudioTOCChapter chapter = reference.audioChapter;
        int position = this.mediaPlayer.getCurrentPosition();
        if (chapter != null) {
            int verseNum = chapter.findVerseByPosition(1, position);
            position = chapter.findPositionOfVerse(verseNum);
        } else {
            position -= 3000;
            position = (position >= 0) ? position : 0;
        }
        AudioPlayState.update(this.controller.activity, reference.getS3Key(), position);
    }

    void advanceToNextItem() {
        if (this.nextReference != null) {
            this.currReference = this.nextReference;
            this.addNextChapter(this.currReference);
            this.preFetchNextChapter(this.currReference);
        } else {
            this.stop();
            AudioPlayState.clear(this.controller.activity); // Must be after stop, because stop does update
        }
    }

    private void addNextChapter(AudioReference reference) {
        if (this.mediaPlayer.isPlaying()) {
            this.mediaPlayer.stop();
            this.mediaPlayer = this.nextPlayer;
            this.mediaPlayer.start();
        } else {
            this.mediaPlayer = this.nextPlayer;
        }
        this.controller.playHasStarted(this.mediaPlayer); // Why is this here it seems redundant?
        this.readVerseMetaData(reference);
    }

    private void preFetchNextChapter(AudioReference reference) {
        this.nextReference = reference.nextChapter();
        if (this.nextReference != null) {
            PreFetchCompletion handler = new PreFetchCompletion();
            AwsS3Cache.shared().readFile(this.nextReference.getS3Bucket(),
                    this.nextReference.getS3Key(),
                    Integer.MAX_VALUE,
                    handler);
        } else {
            nextPlayer = null;
        }
    }
    class PreFetchCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                nextPlayer = initPlayer(file);
                mediaPlayer.setNextMediaPlayer(nextPlayer);
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "NextReadFile Failed " + exception.toString());
        }
    }

    private void readVerseMetaData(AudioReference reference) {
        AudioTOCBible reader = new AudioTOCBible(this.controller.activity, reference.textVersion(), reference.silLang());
        AudioTOCChapter chapter = reader.readVerseAudio(reference.damId(), reference.bookId(), reference.chapterNum());
        reference.audioChapter = chapter;
        if (chapter == null) {
           Log.d(TAG, "Unable to read verse position data");
        }
    }
}

File: Plugins/AudioPlayer/src/ios/AudioPlayer/AudioBible.swift
//
//  AudioBible.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 7/31/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import AVFoundation
import AWS
import Utility

class AudioBible {
    
    private static var instance: AudioBible?
    static func shared(controller: AudioBibleController) -> AudioBible {
        if (AudioBible.instance == nil) {
            AudioBible.instance = AudioBible(controller: controller)
        }
        return AudioBible.instance!
    }
    
    private let controller: AudioBibleController
    private let mediaPlayState: MediaPlayState
    private let controlCenter: AudioControlCenter
    private var audioAnalytics: AudioFauxAnalytics?
    private var player: AVPlayer?
    // Transient Variables
    private var currReference: AudioReference?
    private var nextReference: AudioReference?
    
    private init(controller: AudioBibleController) {
        self.controller = controller
        self.mediaPlayState = MediaPlayState.audio
        self.controlCenter = AudioControlCenter.shared
        self.controlCenter.setupControlCenter(player: self)
        self.initNotifications()
        print("***** Init AudioBible *****")
    }
    
    deinit {
        print("***** Deinit AudioBible *****")
    }
    
    func getPlayer() -> AVPlayer? {
        return self.player
    }
    
    func getCurrentReference() -> AudioReference? {
        return self.currReference
    }
    
    func isPlaying() -> Bool {
        return (self.player != nil) ? self.player!.rate > 0.0 : false
    }
    
    func beginReadFile(reference: AudioReference) {
        print("BibleReader.BEGIN Read File")
        self.currReference = reference
        self.audioAnalytics = AudioFauxAnalytics(mediaSource: reference.tocAudioBook.testament.bible.mediaSource,
                                             mediaId: reference.damId,
                                             languageId: reference.iso3,
                                             textVersion: reference.bibleId,
                                             silLang: reference.iso3)
        self.readVerseMetaData(reference: reference)
        print("INSIDE BibleReader \(reference.damId)")
        self.mediaPlayState.retrieve(mediaId: reference.damId)
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                   s3Key: reference.getS3Key(),
                   expireInterval: Double.infinity,
                   getComplete: { [unowned self] url in
                    if let audioURL = url {
                        self.initAudio(url: audioURL, reference: reference)
                    }
        })
    }
    
    private func initAudio(url: URL, reference: AudioReference) {
        let asset = AVAsset(url: url)
        let playerItem = AVPlayerItem(asset: asset)
        print("Player Item Status \(playerItem.status)")
        
        let seekTime = backupSeek(state: mediaPlayState, reference: reference)
        if (CMTimeGetSeconds(seekTime) > 0.1) {
            playerItem.seek(to: seekTime, completionHandler: nil) // nil handler added 6/2/18
        }
        self.player = AVPlayer(playerItem: playerItem)
        self.player?.actionAtItemEnd = AVPlayer.ActionAtItemEnd.none
        
        self.controller.audioReadyToPlay(enabled: true)
        self.preFetchNextChapter(reference: reference)
    }
    
    private func backupSeek(state: MediaPlayState, reference: AudioReference) -> CMTime {
        if (state.mediaUrl == reference.toString()) {
            if reference.audioChapter != nil {
                return state.position
            } else {
                let duration: Int64 = Int64(Date().timeIntervalSince(state.timestamp))
                let backupSec: Int = String(duration).count // could multiply by a factor here
                let backupTime: CMTime = CMTimeMake(value: Int64(backupSec), timescale: 1)
                return(CMTimeSubtract(state.position, backupTime))
            }
        } else {
            return CMTime.zero
        }
    }
    
    func play() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PLAY *************")
                play.play()
                self.audioAnalytics?.playStarted(item: reference.toString(), position: play.currentTime())
                self.controlCenter.nowPlaying(player: self)
            }
        }
    }
    
    func pause() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PAUSE *************")
                play.pause()
                self.audioAnalytics?.playEnded(item: reference.toString(), position: play.currentTime())
                self.updateMediaPlayStateTime(reference: reference)
            }
        }
    }
    
    func stop() {
        if self.player != nil && self.player!.rate > 0.0 {
            self.pause()
        }
    }
    
    /** This method is called when the prior chapter completes, and by AudioControlCenter.swift
     * when the user clicks the next button.
     */
    func nextChapter() {
        if let play = self.player {
            let startPlayRate = play.rate
            if let curr = self.nextReference {
                self.currReference = curr
                self.addNextChapter(reference: curr)
                NotificationCenter.default.post(name: AudioBibleController.AUDIO_CHAP_CHANGED, object: 1)
                if startPlayRate == 0 {
                    self.audioAnalytics?.playStarted(item: curr.toString(), position: play.currentTime())
                }
            } else {
                self.stop()
                self.mediaPlayState.delete() // Must do after stop, because stop updates
            }
        }
    }
    
    /** This method is called by AudioControlCenter.swift when user clicks the prior button. */
    func priorChapter() {
        if let play = self.player {
            let startPlayRate = play.rate
            if let item = play.currentItem {
                if item.currentTime().seconds < 1.0 {
                    if let prior = self.currReference?.priorChapter() {
                        self.nextReference = self.currReference
                        self.currReference = prior
                        self.addNextChapter(reference: prior)
                        NotificationCenter.default.post(name: AudioBibleController.AUDIO_CHAP_CHANGED,
                                                        object: -1)
                        if startPlayRate == 0 {
                            self.audioAnalytics?.playStarted(item: prior.toString(), position: play.currentTime())
                        }
                    }
                } else {
                    item.seek(to: CMTime.zero, completionHandler: nil) // nil handler added 7/2/18
                    self.controlCenter.nowPlaying(player: self)
                    if startPlayRate == 0 {
                        if let curr = self.currReference {
                            self.audioAnalytics?.playStarted(item: curr.toString(), position: CMTime.zero)
                        }
                    }
                }
            }
            //play.currentItem?.seek(to: kCMTimeZero)
        }
    }
    
    private func initNotifications() {
        let notify = NotificationCenter.default
        notify.addObserver(self,
                           selector: #selector(playerItemDidPlayToEndTime(note:)),
                           name: .AVPlayerItemDidPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemFailedToPlayToEndTime(note:)),
                           name: .AVPlayerItemFailedToPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemPlaybackStalled(note:)),
                           name: .AVPlayerItemPlaybackStalled,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemNewErrorLogEntry(note:)),
                           name: .AVPlayerItemNewErrorLogEntry,
                           object: nil)
        /*notify.addObserver(self,
                            selector: #selector(playerItemNewAccessLogEntry(note:)),
                            name: .AVPlayerItemNewAccessLogEntry,
                            object: nil)*/
        /*notify.addObserver(self,
                            selector: #selector(playerItemTimeJumped(note:)),
                            name: .AVPlayerItemTimeJumped,
                            object: nil)*/
        notify.addObserver(self,
                           selector: #selector(applicationDidFinishLaunching(note:)),
                           name: UIApplication.didFinishLaunchingNotification,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillEnterForeground(note:)),
                           name: UIApplication.willEnterForegroundNotification,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationDidEnterBackground(note:)),
                           name: UIApplication.didEnterBackgroundNotification,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillTerminate(note:)),
                           name: UIApplication.willTerminateNotification,
                           object: nil)
    }

    @objc private func playerItemDidPlayToEndTime(note:Notification) {
        print("\n** DID PLAY TO END \(String(describing: note.object))")
        self.nextChapter()
    }
    @objc private func playerItemFailedToPlayToEndTime(note:Notification) {
        print("\n********* FAILED TO PLAY TO END *********\(String(describing: note.object))")
    }
    @objc private func playerItemPlaybackStalled(note:Notification) {
        print("\n****** PLAYBACK STALLED \(String(describing: note.object))")
    }
    @objc private func playerItemNewErrorLogEntry(note:Notification) {
        print("\n****** ERROR LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.errorLog()))")
    }
    @objc private func playerItemNewAccessLogEntry(note:Notification) {
        print("\n****** ACCESS LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.accessLog()))")
    }
    @objc private func playerItemTimeJumped(note:Notification) {
        print("\n****** TIME JUMPED \(String(describing: note.object))")
    }
    
    /**
    * This method is called when an App is first, launched, not when it is restarted in foreground
    */
    @objc private func applicationDidFinishLaunching(note:Notification) {
        print("\n****** APP DID FINISH LAUNCHING \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when an already running App, which went to background, restarts in foreground.
    */
    @objc private func applicationWillEnterForeground(note:Notification) {
        print("\n****** APP WILL ENTER FOREGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the Home button or Lock button terminate the foreground App
    */
    @objc private func applicationDidEnterBackground(note:Notification) {
        print("\n****** APP DID ENTER BACKGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the App is fully killed.  It is not affected by the Home or Lock button
    */
    @objc private func applicationWillTerminate(note:Notification) {
        print("\n****** APP WILL TERMINATE \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
        self.stop()
    }
    
    private func updateMediaPlayStateTime(reference: AudioReference) {
        var result: CMTime = CMTime.zero
        if let currentTime = self.player?.currentTime() {
            if let audioChapter = reference.audioChapter {
                let verse: Int = audioChapter.findVerseByPosition(priorVerse: 1, time: currentTime)
                result = audioChapter.findPositionOfVerse(verse: verse)
            } else {
                result = currentTime
            }
        }
        self.mediaPlayState.update(mediaUrl: reference.toString(), position: result)
    }
    
    private func addNextChapter(reference: AudioReference) {
        self.player?.pause()
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                                   s3Key: reference.getS3Key(),
                                   expireInterval: Double.infinity,
                                   getComplete: { [unowned self] url in
                                    if let audioURL = url {
                                        let asset = AVAsset(url: audioURL)
                                        let playerItem = AVPlayerItem(asset: asset)
                                        self.player?.replaceCurrentItem(with: playerItem)
                                        self.player?.play()
                                        self.readVerseMetaData(reference: reference)
                                        self.controlCenter.nowPlaying(player: self)
                                    }
                                    self.preFetchNextChapter(reference: reference)
        })
    }
    
    private func preFetchNextChapter(reference: AudioReference) {
        self.nextReference = reference.nextChapter()
        if let next = self.nextReference {
            print("Prefetch \(next.toString())")
            AwsS3Cache.shared.readFile(s3Bucket: next.getS3Bucket(),
                                       s3Key: next.getS3Key(),
                                       expireInterval: Double.infinity,
                                       getComplete: { url in })
        }
    }
    
    private func readVerseMetaData(reference: AudioReference) {
        reference.audioChapter = self.controller.audioTOCBible?.readVerseAudio(
            damid: reference.damId, bookId: reference.bookId, chapter: reference.chapterNum)
    }
}

